### PR TITLE
Add deny grant capability to storages

### PR DESCRIPTION
**Pull request by labkode**

### CODE REVIEW

File: internal/grpc/services/storageprovider/storageprovider.go [hunk 1]
```diff
@@ -59,6 +61,7 @@ type config struct {
 	ExposeDataServer bool                              `mapstructure:"expose_data_server" docs:"false;Whether to expose data server."` // if true the client will be able to upload/download directly to it
 	AvailableXS      map[string]uint32                 `mapstructure:"available_checksums" docs:"nil;List of available checksums."`
 	MimeTypes        map[string]string                 `mapstructure:"mimetypes" docs:"nil;List of supported mime types and corresponding file extensions."`
+	GatewaySvc       string                            `mapstructure:"gatewaysvc" docs:"/;Stores the endpoint at which the GRPC gateway is exposed."`
```
Comment (labkode) [inline internal/grpc/services/storageprovider/storageprovider.go:64]: not needed

File: internal/grpc/services/storageprovider/storageprovider.go [hunk 2]
```diff
@@ -91,6 +94,9 @@ func (c *config) init() {
 	if len(c.AvailableXS) == 0 {
 		c.AvailableXS = map[string]uint32{"md5": 100, "unset": 1000}
 	}
+
+	// get default GatewaySVC value if not set
+	c.GatewaySvc = sharedconf.GetGatewaySVC(c.GatewaySvc)
```
Comment (labkode) [inline internal/grpc/services/storageprovider/storageprovider.go:99]: remove

File: .golangci.yaml [hunk 3]
```diff
@@ -30,7 +30,6 @@ linters:
     - goimports
     - unconvert
     - scopelint
-    - maligned
```
Comment (ishank011) [inline .golangci.yaml:33]: Can we add the maligned check back? The repo has been archived but it still works

File: internal/grpc/services/gateway/usershareprovider.go [hunk 4]
```diff
@@ -399,6 +415,36 @@ func (s *svc) createReference(ctx context.Context, resourceID *provider.Resource
...
+	if err != nil {
+		if _, ok := err.(errtypes.IsNotFound); ok {
+			return status.NewNotFound(ctx, "storage provider not found"), nil
+		}
+		return status.NewInternal(ctx, err, "error finding storage provider"), nil
+	}
+
+	grantRes, err := c.DenyGrant(ctx, grantReq)
+	if err != nil {
+		return nil, errors.Wrap(err, "gateway: error calling AddGrant")
```
Comment (ishank011) [inline internal/grpc/services/gateway/usershareprovider.go:438]: `gateway: error calling DenyGrant`

File: pkg/storage/utils/eosfs/eosfs.go [hunk 5]
```diff
@@ -520,36 +520,88 @@ func (fs *eosfs) UnsetArbitraryMetadata(ctx context.Context, ref *provider.Refer
...
 	}
 
-	rootAuth, err := fs.getRootAuth(ctx)
+	err = fs.c.AddACL(ctx, auth, rootAuth, fn, position, eosACL)
 	if err != nil {
 		return err
 	}
 
-	err = fs.c.AddACL(ctx, auth, rootAuth, fn, eosACL)
+	err = fs.c.AddACL(ctx, auth, rootAuth, fn, position, eosACL)
```
Comment (ishank011) [inline pkg/storage/utils/eosfs/eosfs.go:601]: AddACL is being called twice

File: pkg/eosclient/eosbinary/eosbinary.go [hunk 6]
```diff
@@ -511,6 +529,39 @@ func (c *Client) UnsetAttr(ctx context.Context, auth eosclient.Authorization, at
 	return nil
 }
 
+// GetAttr returns the attribute specified by key
+func (c *Client) GetAttr(ctx context.Context, auth eosclient.Authorization, key, path string) (*eosclient.Attribute, error) {
```
Comment (ishank011) [inline pkg/eosclient/eosbinary/eosbinary.go:533]: This method needs to be added to the `EOSClient` interface

File: pkg/storage/utils/eosfs/eosfs.go [hunk 7]
```diff
@@ -520,36 +520,88 @@ func (fs *eosfs) UnsetArbitraryMetadata(ctx context.Context, ref *provider.Refer
...
+
+	err = fs.c.AddACL(ctx, auth, rootAuth, fn, position, eosACL)
+	if err != nil {
+		return errors.Wrap(err, "eosfs: error adding acl")
+	}
+	return nil
+
+}
+
+func (fs *eosfs) DenyGrant(ctx context.Context, ref *provider.Reference, g *provider.Grantee) error {
```
Comment (ishank011) [inline pkg/storage/utils/eosfs/eosfs.go:560]: We also need to check if the caller has the permission to deny access to the resource by checking the `DenyGrant` bit

File: pkg/storage/utils/grants/grants.go [hunk 8]
```diff
@@ -126,3 +132,48 @@ func GetGranteeType(aclType string) provider.GranteeType {
...
+	return g1 != nil && g2 != nil && cmp.Equal(*g1, *g2)
+}
+
+// IsDenial return true if the grant is a denial grant
+func IsDenial(grant *provider.Grant) bool {
+	return PermissionsEqual(grant.Permissions, &provider.ResourcePermissions{})
+}
+
+// AddGrant adds the newGrant into the list of grants
+func AddGrant(grants *[]*provider.Grant, newGrant *provider.Grant) {
```
Comment (ishank011) [inline pkg/storage/utils/grants/grants.go:152]: Are these methods being used anywhere? Can they be removed?

File: pkg/group/utils/utils.go [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+package utils
+
+import (
+	grouppb "github.com/cs3org/go-cs3apis/cs3/identity/group/v1beta1"
+	"github.com/google/go-cmp/cmp"
+)
+
+// NewEmptyGroup creates an empty group
+func NewEmptyGroup() *grouppb.Group {
```
Comment (ishank011) [inline pkg/group/utils/utils.go:27]: Can we move these to `pkg/utils`?

**Comment by labkode** [comment]
> Isn't this `pkg/storage/utils/grants/grants_test.go` file the one written by Gianmaria to extend the tests?

Those are not needed anymore as they were there only for ensuring correct position but since we have the EOS implementation now these tests (that GIanmaria added) are enough:
pkg/storage/utils/eosfs/eosfs_test.go

File: pkg/storage/utils/grants/grants.go [hunk 10]
```diff
@@ -126,3 +132,13 @@ func GetGranteeType(aclType string) provider.GranteeType {
 		return provider.GranteeType_GRANTEE_TYPE_INVALID
 	}
 }
+
+// PermissionsEqual returns true if the permissions are equal
```
Comment (ishank011) [inline pkg/storage/utils/grants/grants.go:136]: More dead code

File: pkg/eosclient/eosclient.go [hunk 11]
```diff
@@ -27,7 +27,7 @@ import (
 
 // EOSClient is the interface which enables access to EOS instances through various interfaces.
 type EOSClient interface {
-	AddACL(ctx context.Context, auth, rootAuth Authorization, path string, a *acl.Entry) error
+	AddACL(ctx context.Context, auth, rootAuth Authorization, path string, position uint, a *acl.Entry) error
```
Comment (ishank011) [inline pkg/eosclient/eosclient.go:30]: We should add the position parameter to `UpdateACL` as well

**Review by ishank011 — COMMENTED** [review]
Mostly looks good. Couple of minor comments

File: pkg/storage/utils/grants/grants.go (hunk 10 above)
Comment (labkode) [reply pkg/storage/utils/grants/grants.go:136]: This one is used:
```
./internal/grpc/services/gateway/usershareprovider.go:		if grants.PermissionsEqual(req.Grant.Permissions.Permissions, &provider.ResourcePermissions{}) {
```

File: pkg/eosclient/eosclient.go (hunk 11 above)
Comment (labkode) [reply pkg/eosclient/eosclient.go:30]: @ishank011 I've added the parameter, in the future we may consider to remove the UpdateACL call as it not used.

**Review by ishank011 — APPROVED** [review]

**Pull request merged**